framework/Optimizers/GradientDescent.py

```diff
@@ -0,0 +1,840 @@
+    terminate = InputData.parameterInputFactory('terminateFollowers', contentType=InputTypes.BoolType,
+        descr=r"""indicates whether a trajectory should be terminated when it begins following the path
+              of another trajectory.""")
+    terminate.addParam('proximity', param_type=InputTypes.FloatType, required=False,
+        descr=r"""provides the normalized distance at which a trajectory's head should be proximal to
+              another trajectory's path before terminating the following trajectory.""")
+    conv.addSub(terminate)
+    # NOTE to add new convergence options, add them to convergenceOptions above, not here!
+
+    return specs
+
+  @classmethod
+  def getSolutionExportVariableNames(cls):
+    """
+      Compiles a list of acceptable SolutionExport variable options.
+      @ In, None
+      @ Out, ok, dict, {varName: description} for valid solution export variable names
+    """
+    # cannot be determined before run-time due to variables and prefixes.
+    ok = super(GradientDescent, cls).getSolutionExportVariableNames()
+    new = {'stepSize': 'the size of step taken in the normalized input space to arrive at each optimal point'}
+    new['conv_{CONV}'] = 'status of each given convergence criteria'
+    # TODO need to include StepManipulators and GradientApproximators solution export entries as well!
+    # # -> but really should only include active ones, not all of them. This seems like it should work
+    # #    when the InputData can scan forward to determine which entities are actually used.
+    for grad in gradKnownTypes():
+      new.update(gradReturnClass(grad, cls).getSolutionExportVariableNames())
+    for step in stepKnownTypes():
+      new.update(stepReturnClass(step, cls).getSolutionExportVariableNames())
+
+    ok.update(new)
+    return ok
+
+  def __init__(self):
+    """
+      Constructor.
+      @ In, None
+      @ Out, None
+    """
+    RavenSampled.__init__(self)
+    ## Instance Variable Initialization
+    # public
+    self.type = 'GradientDescent Optimizer'
+    # _protected
+    self._gradientInstance = None  # instance of GradientApproximater
+    self._stepInstance = None      # instance of StepManipulator
+    self._acceotInstance = None    # instance of AcceptanceCondition
+    self._gradProximity = 0.01     # TODO user input, the proximity for gradient evaluations
+    # history trackers, by traj, are deques (-1 is most recent)
+    self._gradHistory = {}         # gradients
+    self._stepHistory = {}         # {'magnitude': size, 'versor': direction, 'info': dict} for step
+    self._acceptHistory = {}       # acceptability
+    self._stepRecommendations = {} # by traj, if a 'cut' or 'grow' is recommended else None
+    self._acceptRerun = {}         # by traj, if True then override accept for point rerun
+    self._convergenceCriteria = defaultdict(giveZero) # names and values for convergence checks
```

Comment:
Yes, because we dynamically fill the list of convergence options, a defaultdict that provides a 0 value is handy. Usually I'd use a lambda function, but since those aren't friendly to serialization, I did this instead.

It occurs to me now that perhaps I could always use a "getter" that defaults to 0 instead, as long as developers always know to do that when they check on convergence. Since this is isolated to the GradientDescent, that's probably not too bad. Let me see if that will work.

Reply:
It turns out this would require a somewhat more extensive conditional checking; I feel like this implementation is the cleanest approach.

Because the convergence dictionary is a dictionary of functions, I need a function to call as the default "getter" argument, not a float.